libraries/botbuilder-dialogs/src/memory/scopes/settingsMemoryScope.ts

```diff
@@ -25,16 +35,110 @@ export class SettingsMemoryScope extends MemoryScope {
      * @param dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) object for this turn.
      * @returns The memory for the scope.
      */
-    public getMemory(dc: DialogContext): object {
-        let settings: object = {};
+    public getMemory(dc: DialogContext): Record<string, unknown> {
+        let settings: Record<string, unknown> = {};
         if (dc.context.turnState.has(ScopePath.settings)) {
             settings = dc.context.turnState.get(ScopePath.settings);
-        }
-        for (const key in process.env) {
-            if (typeof process.env[key] == 'string') {
-                settings[key] = process.env[key];
+        } else {
+            const configuration = dc.context.turnState.get(DialogTurnStateConstants.configuration);
+            if (configuration) {
+                settings = SettingsMemoryScope.loadSettings(configuration);
+                dc.context.turnState.set(ScopePath.settings, settings);
             }
         }
+        const envSettings = SettingsMemoryScope.loadSettings(process.env);
+        merge(settings, envSettings);
         return settings;
     }
+
+    /**
+     * Build a dictionary view of configuration providers.
+     *
+     * @param configuration The configuration that we are running with.
+     * @returns Projected dictionary for settings.
+     */
+    protected static loadSettings(configuration: Record<string, string>): Record<string, unknown> {
+        const settings = {};
+        if (configuration) {
+            // load configuration into settings
+            const root = this.convertFlattenSettingToNode(Object.entries(configuration));
+            root.children.forEach((u) => {
+                settings[u.value] = this.convertNodeToObject(u);
+            });
+        }
+        return settings;
+    }
+
+    private static convertFlattenSettingToNode(kvs: [string, string][]): Node {
```

Comment:
These kinds of types read cleaner as `Array<[key: string, value: string]>` to me (I added keys to the tuple, which IIRC is supported in the Typescript version we're using).

I'm also having a bit of trouble visualizing this code - can you write a (non-JSDoc) comment just explaining what it does?

```typescript
// Here is an explanation of this method
private static convertFlattenSerttingToNode(...)
```

Reply:
Added (copied from C#) visualized docs to the code.
Pasted a copy here:
```
/// Generate a node tree with the flatten settings.
/// For example:
/// {
///   "array":["item1", "item2"],
///   "object":{"array":["item1"], "2":"numberkey"}
/// }
/// 
/// Would generate a flatten settings like:
/// array:0 item1
/// array:1 item2
/// object:array:0 item1
/// object:2 numberkey
/// 
/// After Converting it from flatten settings into node tree, would get:
/// 
///                         null
///                |                     |
///              array                object
///            |        |            |        |
///           0          1        array        2
///           |          |         |           |
///         item1       item2      0        numberkey
///                                |
///                              item1
/// The result is a Tree.
```